File: python/semanticabi/steps/TokenTransferStep.py
from typing import List, Dict, Tuple

from semanticabi.abi.SemanticAbi import SemanticAbi
from semanticabi.abi.decoded.TokenTransferDecoded import TokenTransferDecoded
from semanticabi.abi.item.SemanticAbiItem import SemanticAbiItem
from semanticabi.common.column.NumericDatasetColumn import NumericDatasetColumn, NumericType
from semanticabi.common.column.StringDatasetColumn import StringType
from semanticabi.metadata.EthBlock import EthBlock
from semanticabi.metadata.EthTransaction import EthTransaction
from semanticabi.steps.AbiSchema import AbiSchema
from semanticabi.steps.Step import Step, TransformItem


_SCHEMA: AbiSchema = AbiSchema([
    StringType.ADDRESS_HASH('fromAddress'),
    StringType.ADDRESS_HASH('toAddress'),
    NumericDatasetColumn.int256('value', higher_order_type=NumericType.CURRENCY),
    StringType.ID('tokenId'),
    StringType.ENUM('tokenType')
])


class TokenTransferTransformItem(TransformItem):
    _token_transfer: TokenTransferDecoded

    def __init__(self, token_transfer: TokenTransferDecoded):
        # We don't have a DecodedResult that we need to pass down
        super().__init__(None)
        self._token_transfer = token_transfer

    @property
    def contract_address(self) -> str:
        return self._token_transfer.contract_address

    @property
    def internal_index(self) -> str:
        return str(self._token_transfer.internal_index)

    @property
    def item_type(self) -> str:
        return 'transfer'


class TokenTransferStep(Step):
    """
    Special step to handle token transfer matches. This takes all the various transfer events from the transaction and
    generates a single row for each transfer event, with the from, to, value, tokenId, and tokenType fields
    """

    @property
    def _abi(self) -> SemanticAbi:
        raise Exception('Method not implemented.')

    @property
    def _abi_item(self) -> SemanticAbiItem:
        raise Exception('Method not implemented.')

    @property
    def schema(self) -> AbiSchema:
        return _SCHEMA

    def _inner_transform(self, block: EthBlock, transaction: EthTransaction) -> List[Tuple[TransformItem, List[Dict[str, any]]]]:
        results: List[Tuple[TransformItem, List[Dict[str, any]]]] = []

        for transfer in transaction.transfers:
            results.append((TokenTransferTransformItem(transfer), [{
                'fromAddress': transfer.from_address,
                'toAddress': transfer.to_address,
                'value': transfer.value,
                'tokenId': str(transfer.token_id),
                'tokenType': transfer.token_type.code
            }]))

        return results

File: js/src/semanticabi/transform/steps/TokenTransferStep.ts
import {Step} from "semanticabi/transform/steps/Step";
import {AbiSchema} from "semanticabi/transform/steps/AbiSchema";
import {DataType} from "semanticabi/transform/column/DataType";
import {TokenTransferDecoded} from "semanticabi/metadata/TokenTransferDecoded";
import {TransformItem} from "semanticabi/transform/steps/TransformItem";
import {EthBlock} from "semanticabi/metadata/EthBlock";
import {EthTransaction} from "semanticabi/metadata/EthTransaction";
import {JsonObject} from "common/CommonTypes";
import { SemanticAbiItem } from "semanticabi/abi/item/semantic/SemanticAbiItem";
import { SemanticAbi } from "semanticabi/abi/SemanticAbi";

const SCHEMA: AbiSchema = new AbiSchema([
    DataType.hashColumn('fromAddress'),
    DataType.hashColumn('toAddress'),
    DataType.NUMBER.column('value'),
    DataType.STRING.column('tokenId'),
    DataType.STRING.column('tokenType')
]);

/**
 * Special step to handle token transfer matches. This takes all the various transfer events from the transaction and
 * generates a single row for each transfer event, with the from, to, value, tokenId, and tokenType fields.
 */
export class TokenTransferStep extends Step {

    get abi(): SemanticAbi {
        throw new Error("Method not implemented.");
    }

    get abiItem(): SemanticAbiItem {
        throw new Error("Method not implemented.");
    }

    get shouldTransform(): boolean {
        throw new Error("Method not implemented.");
    }

    get schema(): AbiSchema {
        return SCHEMA;
    }

    innerTransform(block: EthBlock, transaction: EthTransaction): [TransformItem, JsonObject[]][] {
        return transaction.transfers.map(transfer =>
            [
                new TokenTransferTransformItem(transfer),
                [{
                    'fromAddress': transfer.fromAddress,
                    'toAddress': transfer.toAddress,
                    'value': transfer.value.nullable,
                    'tokenId': transfer.tokenId.map(id => id.toString()).nullable,
                    'tokenType': transfer.tokenType.name
                }]
            ]
        );
    }

}

class TokenTransferTransformItem extends TransformItem {

    constructor(
        private readonly tokenTransfer: TokenTransferDecoded
    ) {
        super(() => null);
    }

    get contractAddress(): string {
        return this.tokenTransfer.contractAddress;
    }

    get internalIndex(): string {
        return this.tokenTransfer.internalIndex.toString();
    }

    get itemType(): string {
        return 'transfer';
    }

}
